kqueen/engines/heat.py

```diff
@@ -0,0 +1,178 @@
+from collections import namedtuple
+import importlib
+import yaml
+import logging
+import sys
+import os
+
+from kqueen.config import current_config
+import openstack
+
+from .base import BaseEngine
+
+logger = logging.getLogger('kqueen_api')
+config = current_config()
+
+
+STATE_MAP = {
+    'IN_PROGRESS': config.get('CLUSTER_PROVISIONING_STATE'),
+    'COMPLETE': config.get('CLUSTER_OK_STATE'),
+}
+
+
+class Heat:
+    templates = {
+        "default": "kqueen.engines.resources.heat.default",
+        "empty": "kqueen.engines.resources.heat.empty",
+    }
+
+    def __init__(self, **kwargs):
+        self.connection = openstack.connection.Connection(**kwargs)
+
+    def _get_template_and_files(self, name):
+        # NOTE(sskripnick)
+        # we need to custom load and decode all files because of this bug:
+        # https://storyboard.openstack.org/#!/story/2002002
+        package = importlib.import_module(self.templates[name])
+        path = list(package.__path__)[0]
+        files = {}
+        for root, dirs, _files in os.walk(path):
+            for f in _files:
+                fpath = os.path.join(root, f)
+                files["file://" + fpath] = open(fpath, "rb").read().decode("utf8")
+        return os.path.join(path, "template.yaml"), files
+
+    def create(self, *, name, template_name, ssh_key_name, server_image):
+        template_file, files = self._get_template_and_files(template_name)
+        try:
+            self.connection.create_stack(
+                name=name,
+                template_file=template_file,
+                files=files,
+                ssh_key_name=ssh_key_name,
+                server_image=server_image,
+            )
+        except openstack.exceptions.BadRequestException as e:
+            # NOTE(sskripnick)
+            # handle this exception separately because useful information
+            # may be found only in e.details
+            logger.exception("Failed to create stack: %s" % e.details)
```

Comment:
use '.format()' directive

Reply:
+ 1

Reply:
Why?

Reply:
https://docs.python.org/3.1/library/string.html#format-examples
It is officially old

Reply:
Starting from python 3.6 we can use this https://docs.python.org/3.6/reference/lexical_analysis.html#f-strings
E.g. `"Failed to create stack: {e.details}"`

Reply:
@katyafervent what is officially old?

Reply:
"%s" symbol architecture is deprecated

Reply:
@naumvd95 never heard about it. Any links?